hoomd/mpcd/RadialSolventVelocityAnalyzer.cc

```diff
@@ -0,0 +1,184 @@
+#include "hoomd/extern/pybind/include/pybind11/stl.h"
+
+namespace mpcd
+{
+
+/*!
+ * \param sysdata MPCD system data
+ * \param R Cutoff radius for calculating Radial Density profile
+ * \param bin_width Width for binning particle distances
+ */
+RadialSolventVelocityAnalyzer::RadialSolventVelocityAnalyzer(std::shared_ptr<mpcd::SystemData> sysdata,
+                                                             Scalar R,
+                                                             Scalar bin_width)
+    : Analyzer(sysdata->getSystemDefinition()),
+      m_mpcd_sys(sysdata),
+      m_mpcd_pdata(m_mpcd_sys->getParticleData()),
+      m_R(R)
+    {
+    m_exec_conf->msg->notice(5) << "Constructing RadialSolventVelocityAnalyzer" << std::endl;
+
+    assert(m_bin_width > 0.0);
+
+    // allocate memory for the bins
+    m_num_bins = std::round(m_R / bin_width);
+    m_bin_width = m_R / m_num_bins;
+    reset();
+
+#ifdef ENABLE_MPI
+    // MPI is currently not supported due to how this analyzer might degrade performance
+    if (m_exec_conf->getNRanks() > 1)
+        {
+        m_exec_conf->msg->error() << "MPCD: Radial Solvent Velocity analyzer does not support MPI" << std::endl;
+        throw std::runtime_error("Radial Solvent Velocity analyzer does not support MPI");
+        }
+#endif // ENABLE_MPI
+    }
+
+RadialSolventVelocityAnalyzer::~RadialSolventVelocityAnalyzer()
+    {
+    m_exec_conf->msg->notice(5) << "Destroying RadialSolventVelocityAnalyzer" << std::endl;
+    }
+
+/*!
+ * \param timestep Current simulation timestep
+ *
+ * To get rho(r), we calculate as
+ *
+ * \verbatim
+ *                  counts_i
+ *      g_i = ----------------------
+ *                  V_shell
+ *
+ * \endverbatim
+ *
+ * The shell volume is V_shell = (4 pi/3)*(r_{i+1}^3-r_i^3).
+ */
+void RadialSolventVelocityAnalyzer::analyze(unsigned int timestep)
+    {
+    if (m_prof) m_prof->push("RadialSolventVelocityAnalayse");
+
+    // binParticles
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(), access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(), access_location::host, access_mode::read);
+    std::vector<unsigned int> counts(m_num_bins, 0);
+    std::vector<Scalar> radial_vel(m_num_bins, 0);
+
+    // compute the total number of mpcd particles
+    const unsigned int N = m_mpcd_pdata->getN();
+    for (unsigned int i=0; i < N; ++i)
+        {
+        // load particle i
+        const Scalar4 pos_i = h_pos.data[i];
+        const Scalar4 vel_i = h_vel.data[i];
+        const Scalar3 r_i = make_scalar3(pos_i.x, pos_i.y, pos_i.z);
+        const Scalar3 v_i = make_scalar3(vel_i.x, vel_i.y, vel_i.z);
+        
+        // distance and radial velocity calculation 
+        const Scalar r_isq = dot(r_i, r_i);
+        const Scalar r_i_norm = slow::sqrt(r_isq);
+        const Scalar v_irad = dot(v_i, r_i)/r_i_norm;
+
+        if (r_i_norm < m_R)
+            {
```

Comment:
Micro-optimization: only do the hard calculations for particles inside the requested radius:
```suggestion
        const Scalar4 pos_i = h_pos.data[i];
        const Scalar3 r_i = make_scalar3(pos_i.x, pos_i.y, pos_i.z);
        const Scalar r_isq = dot(r_i, r_i);

        if (r_i_norm < m_R)
            {
            const Scalar4 vel_i = h_vel.data[i];
            const Scalar3 v_i = make_scalar3(vel_i.x, vel_i.y, vel_i.z);
            const Scalar r_i_norm = slow::sqrt(r_isq);
            const Scalar v_irad = dot(v_i, r_i)/r_i_norm;
```